### ISSUE TITLE
Elastic fwi for rebar in concrete

### ISSUE THREAD

**Issue by aaelmeli** [ar4/deepwave#91]
Hi Alan,

I am trying to use elastic FWI to reconstruct rebar embedded in concrete with no success. I have tried different setups and was wondering if I am doing something wrong or you have any thoughts on how to get it to work. Please find the code I use below:

Thanks alot.

```
import torch
import numpy as np
import matplotlib.pyplot as plt
import deepwave
from deepwave import elastic

# Set the physical dimensions of the model
beam_depth_cm = 25  # Depth of the beam in cm
beam_length_cm = 25  # Length of the beam in cm

# Set the grid dimensions
rows = 100  # Number of rows representing the depth
cols = 100  # Number of columns representing the length
dx = beam_length_cm / cols   # Convert grid spacing from cm to meters

# Base wave speed for concrete (in m/s)
base_wave_speed = 4000.  # Example base wave speed

# Rebar wave speed (in m/s)
rebar_wave_speed = 6000.  # Example wave speed for rebar

# Diameter of the circular inclusions (in cm)
inclusion_diameter_cm = 2
inclusion_radius_cm = inclusion_diameter_cm / 2
inclusion_radius_rows = int(inclusion_radius_cm / (beam_depth_cm / rows))
inclusion_radius_cols = int(inclusion_radius_cm / (beam_length_cm / cols))

# Generate the base wave speed model
vp_model = torch.full((rows, cols), base_wave_speed, dtype=torch.float32)

# Center of the beam
beam_center_row = rows // 2
beam_center_col = cols // 2

# Define the positions of the circular inclusions
inclusion_centers = [
    # (beam_center_row, beam_center_col - int(beam_length_cm / 4 / dx)),
    (beam_center_row, beam_center_col),
    # (beam_center_row, beam_center_col + int(beam_length_cm / 4 / dx))
]

# Create the inclusions
for i, (center_row, center_col) in enumerate(inclusion_centers):
    for row in range(rows):
        for col in range(cols):
            distance = np.sqrt((row - center_row)**2 + (col - center_col)**2)
            if distance <= inclusion_radius_rows:
                if i != 1 or (i == 1 and not (col > center_col + 2)):  # Chamfer right side
                    vp_model[row, col] = rebar_wave_speed

# Compute Vs and rho for the concrete
poisson_ratio = 0.25
density = 2500

# Compute Vs and rho
vs_model = vp_model / (2 * (1 + poisson_ratio))
a = 310
b = 0.25
rho_model = a * vp_model**b

# Initial models
vp_initial = torch.full((rows, cols), base_wave_speed, dtype=torch.float32)
vs_initial = vp_initial / (2 * (1 + poisson_ratio))
rho_initial = torch.full((rows, cols), density, dtype=torch.float32)

# Define the finite difference parameters
n_shots = 24
n_sources_per_shot = 1
d_source = 4
first_source = 4
source_depth = 0

n_receivers_per_shot = 96
d_receiver = 1
first_receiver = 1
receiver_depth = 0

dt = 0.5e-6
t = 200e-6  # sec
nt = int(t / dt)
freq = 100e3  # Hz
peak_time = 1.5 / freq
dx = 0.25e-2  

# Device configuration
device = torch.device('cuda')

# Source locations
source_locations = torch.zeros(n_shots, n_sources_per_shot, 2, dtype=torch.long, device=device)
source_locations[..., 0] = source_depth
source_locations[:, 0, 1] = (torch.arange(n_shots) * d_source + first_source)


source_locations2 = torch.zeros(n_shots, n_sources_per_shot, 2,
                               dtype=torch.long, device=device)
source_locations2[..., 0] = 98
source_locations2[:, 0, 1] = (torch.arange(n_shots) * d_source +
                             first_source)

# source_locations=torch.cat((source_locations, source_locations2), dim=0)


# Receiver locations
receiver_locations = torch.zeros(n_shots, n_receivers_per_shot, 2, dtype=torch.long, device=device)
receiver_locations[..., 0] = receiver_depth
receiver_locations[:, :, 1] = (torch.arange(n_receivers_per_shot) * d_receiver + first_receiver).repeat(n_shots, 1)

receiver_locations2 = torch.zeros(n_shots, n_receivers_per_shot, 2,
                                 dtype=torch.long, device=device)
receiver_locations2[..., 0] = 98
receiver_locations2[:, :, 1] = (
    (torch.arange(n_receivers_per_shot) * d_receiver +
     first_receiver)
    .repeat(n_shots, 1)
)

# receiver_locations=torch.cat((receiver_locations, receiver_locations2), dim=0)

# Source amplitudes
source_amplitudes = 1e10 * (deepwave.wavelets.ricker(freq, nt, dt, peak_time).repeat(n_shots, n_sources_per_shot, 1).to(device))

# Compute CFL condition
cfl = base_wave_speed * dt / dx
print("CFL condition: ", cfl)

# Observed data
vp = vp_model.to(device)
vs = vs_model.to(device)
rho = rho_model.to(device)

observed_data = elastic(
    *deepwave.common.vpvsrho_to_lambmubuoyancy(vp, vs, rho),
    dx, dt,
    source_amplitudes_y=source_amplitudes,
    source_locations_y=source_locations,
    receiver_locations_y=receiver_locations,
    pml_width=[0, 0, 0, 0]
)

# Model definition
class Model(torch.nn.Module):
    def __init__(self, initial, min_val, max_val):
        super().__init__()
        self.min_val = min_val
        self.max_val = max_val
        self.model = torch.nn.Parameter(
            torch.logit((initial - min_val) / (max_val - min_val))
        )

    def forward(self):
        return torch.sigmoid(self.model) * (self.max_val - self.min_val) + self.min_val

# Instantiate models
vp_model = Model(vp_initial, vp.min().item(), vp.max().item()).to(device)
vs_model = Model(vs_initial, vs.min().item(), vs.max().item()).to(device)
rho_model = Model(rho_initial, rho.min().item(), rho.max().item()).to(device)



vp_trial = vp_initial.clone().requires_grad_()
vs_trial = vs_initial.clone().requires_grad_()
# rho_trial = rho_initial.clone().requires_grad_()

# Optimizer
optimiser = torch.optim.LBFGS([vp_trial, vs_trial], lr=0.1, line_search_fn='strong_wolfe')

# Loss function
loss_fn = torch.nn.L1Loss()
# loss_fn = torch.nn.MSELoss()

# Function to calculate normalized dot product loss
def normalized_dot_product_loss(pred, target):
    pred_flat = pred.reshape(pred.size(0), -1)
    target_flat = target.reshape(pred.size(0), -1)
    
    dot_product = torch.sum(pred_flat * target_flat, dim=1)
    norm_pred = torch.norm(pred_flat, dim=1)
    norm_target = torch.norm(target_flat, dim=1)
    
    normalized_dot = dot_product / (norm_pred * norm_target + 1e-8)  # Add epsilon to avoid division by zero
    return 1 - torch.mean(normalized_dot)

# Run optimization/inversion

n_epochs = 500

# Function to calculate TV norm
def tv_norm(x):
    if x.dim() == 2:  # 2D tensor
        diff_x = torch.abs(x[:-1, :] - x[1:, :]).sum()
        diff_y = torch.abs(x[:, :-1] - x[:, 1:]).sum()
        return diff_x + diff_y
    elif x.dim() == 3:  # 3D tensor
        diff_x = torch.abs(x[:, :-1, :] - x[:, 1:, :]).sum()
        diff_y = torch.abs(x[:, :, :-1] - x[:, :, 1:]).sum()
        return diff_x + diff_y
    else:
        raise ValueError("Unsupported tensor dimension for TV norm")

lambda_tv = 0.1  # Regularization parameter

for epoch in range(n_epochs):
    def closure():
        optimiser.zero_grad()
        out = elastic(
            *deepwave.common.vpvsrho_to_lambmubuoyancy(vp_trial.to(device), vs_trial.to(device), rho.to(device)),
            dx, dt,
            source_amplitudes_y=source_amplitudes,
            source_locations_y=source_locations,
            receiver_locations_y=receiver_locations,
            pml_width=[0, 0, 0, 0]
        )
        loss = 1e10 * (loss_fn(out[-2], observed_data[-2]))
        # loss = 1e10 *normalized_dot_product_loss(out[-2],observed_data[-2])
        # tv_loss = lambda_tv * (tv_norm(vp_trial) + tv_norm(vs_trial))
        total_loss = loss 
        print(f"Epoch {epoch}, Loss: {total_loss.item()}")
        filen = f'./results/concrete/elastic/L1_TvReg/vp_{epoch}.npy'
        np.save(filen, vp_trial.cpu().detach().numpy())
        filen = f'./results/concrete/elastic/L1_TvReg/vs_{epoch}.npy'
        np.save(filen, vs_trial.cpu().detach().numpy())
        # filen = f'./results/concrete/elastic/L2_TvReg/rho_{epoch}.npy'
        # np.save(filen, rho_trial.cpu().detach().numpy())
        total_loss.backward()
        return total_loss

    optimiser.step(closure)

```

**Comment by ar4** [comment]
Hi Abdo,

Can you provide some more detail about what kind of results you have been
getting?

Some initial thoughts:

1. Have you tried inverting from low to high frequency?
2. I think it is generally only possible to invert down to a depth of about
1/3 the length of the source/receiver array if they are on one surface.
Your anomaly is deeper than that, so FWI will have a hard time. Have you
tried having sources and receivers on opposite sides of the anomaly (e.g.,
sources on the top, receivers on the bottom)?
3. A 2000 m/s anomaly is quite large, so FWI might struggle with that.
4. If you are going to use the Model class, you should have the min and max
a bit lower and higher than the actual min/max, respectively, because it
uses a sigmoid.

**Comment by aaelmeli** [comment]
> Hi Abdo,
> 
> Can you provide some more detail about what kind of results you have been
> getting?
> 
> Some initial thoughts:
> 
> 1. Have you tried inverting from low to high frequency?
> 2. I think it is generally only possible to invert down to a depth of about
> 1/3 the length of the source/receiver array if they are on one surface.
> Your anomaly is deeper than that, so FWI will have a hard time. Have you
> tried having sources and receivers on opposite sides of the anomaly (e.g.,
> sources on the top, receivers on the bottom)?
> 3. A 2000 m/s anomaly is quite large, so FWI might struggle with that.
> 4. If you are going to use the Model class, you should have the min and max
> a bit lower and higher than the actual min/max, respectively, because it
> uses a sigmoid.

Hi Alan, 

Thanks for your reply and suggestions. I could figure it out. I think I was using some empirical equation to obtain the densities that was causing in sufficient sensitivity for the parameters. Once I fixed it, I got the desired results. without the need for multi-resolution.

Thanks again for your effort.
